### PR TITLE
IM test logic fixing

### PR DESCRIPTION
**Pull request by joelanders**
https://github.com/measurement-kit/measurement-kit/issues/1492

This fixes all remaining bugs.

### CODE REVIEW

**Comment by joelanders** [comment]
Bleh I see the CI's compiler wants me to add a declaration for the new "is this IP in FB's ASN?" function I added.

File: src/libmeasurement_kit/ooni/telegram.cpp [hunk 1]
```diff
@@ -34,7 +34,7 @@ static void tcp_many(const std::vector<std::string> ip_ports, SharedPtr<Entry> e
                 logger->info("telegram: success TCP connecting to %s:%d",
                     ip.c_str(), port);
                 result["status"]["success"] = true;
-                (*entry)["telegram_tcp_blocking"] = true;
+                (*entry)["telegram_tcp_blocking"] = false;
```
Comment (hellais) [inline src/libmeasurement_kit/ooni/telegram.cpp]: This is not the only logic bug. You are also re-initialising `telegram_tcp_blocking` to `true` in both `http_many` and `tcp_many`. Doing so means that the condition "consider it to work if at least one is working" will not be met in many cases.

**Comment by hellais** [comment]
> CI's compiler wants me to add a forward declaration for the new "is this IP in FB's ASN?"

@bassosimone knows better, but if the function is defined as not defined as `static` you need to declare it inside of the header. You should either mark it as static or a declaration for it.

**Comment by hellais** [comment]
I haven't actually run it, but a cursory look at the diff looks good.

As mentioned on slack there is still one logic bug missing to be squashed:

> Only populate the whatsapp_endpoints_blocked with the endpoints that are blocked AND consistent

Current you are erasing it from the list of `blocked_hostnames` only when it fails to connect and is consistent: https://github.com/measurement-kit/measurement-kit/blob/48f4e3cefdd60806a14e2c5f8dd438d6e6b5df57/src/libmeasurement_kit/ooni/whatsapp.cpp#L241, you should be doing the opposite.
Erasing it from the list of blocked domains when either it succeeds or it fails but is inconsistent.

**Comment by hellais** [comment]
My checklist for testing this is:

* [ ] Do another pass at reading the diff
* [ ] Smoke test by compiling and running it
* [ ] Setup a local system to simulate various forms of telegram, whatsapp and fb blocking and confirm it produces consistent measurements
* [ ] Merge

**Comment by hellais** [comment]
Here are some things I am noticing while comparing the output of `ooniprobe` to the output of the MK test:

## Telegram

* The telegram test output misses the key `telegram_http_blocking`. Grepping for this key in the source I don't find it.

**Comment by joelanders** [comment]
@hellais ready for another review.

**Comment by hellais** [comment]
@joelanders so it seems like the facebook test is still not writing the keys properly.

Here is a sample of the report having set my `/etc/hosts` with what I described above:

```
        "facebook_b_api_dns_consistent": false,
        "facebook_b_api_reachable": false,
        "facebook_b_graph_dns_consistent": false,
        "facebook_b_graph_reachable": false,
        "facebook_dns_blocking": true,
        "facebook_edge_dns_consistent": true,
        "facebook_edge_reachable": true,
        "facebook_external_cdn_dns_consistent": true,
        "facebook_external_cdn_reachable": true,
        "facebook_scontent_cdn_dns_consistent": true,
        "facebook_scontent_cdn_reachable": true,
        "facebook_star_dns_consistent": true,
        "facebook_star_reachable": true,
        "facebook_stun_dns_consistent": false,
        "facebook_stun_reachable": null,
        "facebook_tcp_blocking": true,
```

It states that there is `facebook_tcp_blocking`, but that is not the case as the DNS response is inconsistent. Similarly to the whatsapp and telegram test you should set `tcp_blocking` to true only when CONSISTENT IPs are unreachable (not inconsistent as in this case).

**Comment by hellais** [comment]
I checked the telegram and whatsapp tests and things there are good 👍

**Comment by joelanders** [comment]
@hellais pong.

**Comment by hellais** [comment]
Ok, I did another pass of testing and this looks good to me.

🐳 

I will let @bassosimone do the honors of merging it if it looks good to him as well.

**Comment by bassosimone** [comment]
Yap, will look into this later today!

File: src/libmeasurement_kit/ooni/telegram.cpp (hunk 1 above)
Comment (bassosimone) [reply src/libmeasurement_kit/ooni/telegram.cpp]: FTR, it seems that now this has been fixed!

**Pull request merged**